core/state/snapshot/difflayer.go

```diff
@@ -0,0 +1,337 @@
+func (dl *diffLayer) Account(hash common.Hash) *Account {
+	data := dl.AccountRLP(hash)
+	if len(data) == 0 { // can be both nil and []byte{}
+		return nil
+	}
+	account := new(Account)
+	if err := rlp.DecodeBytes(data, account); err != nil {
+		panic(err)
+	}
+	return account
+}
+
+// AccountRLP directly retrieves the account RLP associated with a particular
+// hash in the snapshot slim data format.
+func (dl *diffLayer) AccountRLP(hash common.Hash) []byte {
+	dl.lock.RLock()
+	defer dl.lock.RUnlock()
+
+	// If the account is known locally, return it. Note, a nil account means it was
+	// deleted, and is a different notion than an unknown account!
+	if data, ok := dl.accountData[hash]; ok {
+		return data
+	}
+	// Account unknown to this diff, resolve from parent
+	return dl.parent.AccountRLP(hash)
+}
+
+// Storage directly retrieves the storage data associated with a particular hash,
+// within a particular account. If the slot is unknown to this diff, it's parent
+// is consulted.
+func (dl *diffLayer) Storage(accountHash, storageHash common.Hash) []byte {
+	dl.lock.RLock()
+	defer dl.lock.RUnlock()
+
+	// If the account is known locally, try to resolve the slot locally. Note, a nil
+	// account means it was deleted, and is a different notion than an unknown account!
+	if storage, ok := dl.storageData[accountHash]; ok {
+		if storage == nil {
+			return nil
+		}
+		if data, ok := storage[storageHash]; ok {
+			return data
+		}
+	}
+	// Account - or slot within - unknown to this diff, resolve from parent
+	return dl.parent.Storage(accountHash, storageHash)
+}
+
+// Update creates a new layer on top of the existing snapshot diff tree with
+// the specified data items.
+func (dl *diffLayer) Update(blockRoot common.Hash, accounts map[common.Hash][]byte, storage map[common.Hash]map[common.Hash][]byte) *diffLayer {
+	return newDiffLayer(dl, dl.number+1, blockRoot, accounts, storage)
+}
+
+// Cap traverses downwards the diff tree until the number of allowed layers are
+// crossed. All diffs beyond the permitted number are flattened downwards. If
+// the layer limit is reached, memory cap is also enforced (but not before). The
+// block numbers for the disk layer and first diff layer are returned for GC.
+func (dl *diffLayer) Cap(layers int, memory uint64) (uint64, uint64) {
+	// Dive until we run out of layers or reach the persistent database
+	if layers > 2 {
+		// If we still have diff layers below, recurse
+		if parent, ok := dl.parent.(*diffLayer); ok {
+			return parent.Cap(layers-1, memory)
+		}
+		// Diff stack too shallow, return block numbers without modifications
+		return dl.parent.(*diskLayer).number, dl.number
+	}
+	// We're out of layers, flatten anything below, stopping if it's the disk or if
+	// the memory limit is not yet exceeded.
+	switch parent := dl.parent.(type) {
+	case *diskLayer:
+		return parent.number, dl.number
+	case *diffLayer:
+		dl.lock.Lock()
+		defer dl.lock.Unlock()
+
+		dl.parent = parent.flatten()
```

Comment:
We might need some smarter locking here. If the parents have side branches, those don't get locked by the child's lock.